pkg/summarizer/enhancedanalyzer/analyzer.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2020 The TestGrid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// An extension of naiveanalyzer with an improved algorithm
+package enhancedanalyzer
+
+import (
+	statepb "github.com/GoogleCloudPlatform/testgrid/pb/state"
+	summarypb "github.com/GoogleCloudPlatform/testgrid/pb/summary"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer/baseanalyzer"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer/common"
+)
+
+const analyzerName = "enhancedanalyzer"
+
+// IntString is for sorting, primarily intended for map[string]int as implemented below
+type IntString struct {
+	s string
+	i int
+}
+
+// EnhancedAnalyzer implements functions that calculate flakiness as a ratio of failed tests to total tests
+type EnhancedAnalyzer struct {
+}
+
+// GetFlakiness returns a HealthinessInfo message
+func (ea *EnhancedAnalyzer) GetFlakiness(gridMetrics []*common.GridMetrics, relevantFilteredStatus map[string][]statepb.Row_Result, minRuns int, startDate int, endDate int, tab string) *summarypb.HealthinessInfo {
+	// Delegate to a BaseAnalyzer and change the flakiness scores.
+	// (And average flakiness)
+	na := baseanalyzer.BaseAnalyzer{}
+	hi := na.GetFlakiness(gridMetrics, relevantFilteredStatus, minRuns, startDate, endDate, tab)
+	averageFlakiness := float32(0)
+	for _, test := range hi.Tests {
+		test.Flakiness = calculateEnhancedFlakiness(relevantFilteredStatus[test.DisplayName])
+		averageFlakiness += test.Flakiness
+	}
+	hi.AverageFlakiness = averageFlakiness / float32(len(hi.Tests))
+	return hi
+}
+
+const IGNORE_FAILURES_IN_A_ROW = 3
```

Comment:
Is this existing behavior? I know this is supposed to discount 'actual failures' from counting as flakes, but I also feel like this is kinda specific to GKE's use case? Status flips should be able to tell that kind of story on their own, _or_ we should use the configured failure threshold (or whether the row has an 'alert' on it) to determine.

Reply:
This is existing behavior in the GKE-specific autojailer.  It exists to handle a (depressingly common) case where a test is legitimately failing for most of the time period and flaky for the rest of it, and so looks healthy.